daemon: Don't auto disable session affinity

brb opened the pull request:
This commit doesn't auto disable the session affinity feature when KPR
is disabled.

The session affinity doesn't depend on KPR for a ClusterIP service
translation in a pod netns implemented in bpf_lxc (the host netns case
should be covered by kube-proxy's rules). Therefore, we should allow
users to choose whether the feature should be enabled.

Note, that we don't enable the feature by default, as it might increase
bpf_lxc complexity.

Ref #16163.

brb committed:
daemon: Don't auto disable session affinity

This commit doesn't auto disable the session affinity feature when KPR
is disabled.

The session affinity doesn't depend on KPR for a ClusterIP service
translation in a pod netns implemented in bpf_lxc (the host netns case
should be covered by kube-proxy's rules). Therefore, we should allow
users to choose whether the feature should be enabled.

Note, that we don't enable the feature by default, as it might increase
bpf_lxc complexity.

Signed-off-by: Yuan Liu <liuyuan@google.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

pchaigno reviewed (changes requested):
Besides André's comment, I think we'll also want to update the documentation (e.g., https://docs.cilium.io/en/latest/operations/system_requirements/).

borkmann reviewed (approved):
Ok, lgtm. Please also update KPR doc under session affinity section to reflect this new behavior.

brb committed:
docs: Update KPR GSG wrt enabling SA when KPR=disabled

Also, mention the kernel requirements for SA with kube-proxy's
ClusterIP.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

pchaigno commented on Documentation/gettingstarted/kubeproxy-free.rst:
```diff
@@ -1148,6 +1153,13 @@ a fixed cookie value as a trade-off. This makes all applications on the host to
+(until `GH#16197<https://github.com/cilium/cilium/issues/16197>`__ is fixed).
+For this case the session affinity support is disabled by default. To enable the
+feature, set ``config.sessionAffinity=true``.
+
```
I think we also need to document the kernel requirement in https://docs.cilium.io/en/latest/operations/system_requirements/.

pchaigno commented on Documentation/gettingstarted/kubeproxy-free.rst:
```diff
@@ -1136,10 +1136,15 @@ if needed.
+:ref:`host-services` operate (a source IP is not available there, as the
+endpoint selection happens before a network packet has been built by the
+kernel). If :ref:`host-services` is not used (i.e. the loadbalancing is done by
+the BPF program ``bpf_lxc`` attached to a pod network interface), then the
```
Maybe we can avoid references to BPF program names in a documentation destined to users?
```suggestion
kernel). If :ref:`host-services` is not used (i.e. the loadbalancing is done at
the pod network interface, on a per-packet basis), then the
```

pchaigno commented on Documentation/gettingstarted/kubeproxy-free.rst:
```diff
@@ -1148,6 +1153,13 @@ a fixed cookie value as a trade-off. This makes all applications on the host to
 
+For users who run with kube-proxy (i.e. with Cilium's kube-proxy replacement
+disabled), the ClusterIP service loadbalancing when a request is sent from a pod
+running in a non-host network namespace is still performed by the ``bpf_lxc``
```
Ditto.
```suggestion
running in a non-host network namespace is still performed at the pod network interface
```

pchaigno commented on test/k8sT/Services.go:
```diff
@@ -1979,6 +1982,10 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
 		})
 
+		SkipItIf(helpers.DoesNotRunOn419OrLaterKernel, "Tests ClusterIP with sessionAffinity from pods", func() {
+			testSessionAffinity(false, true)
```
How is that different from the existing `Tests NodePort with sessionAffinity`? Maybe we need to redeploy Cilium with KPR disabled here?

brb replied:
Extended the skip condition to make it more obvious that it should run with KPR=disabled.

pchaigno replied:
But when is that condition ever true? `helpers.RunsOn419OrLaterKernel() && helpers.RunsWithoutKubeProxyReplacement()`

brb replied:
On GKE it should be true.

pchaigno replied:
It isn't: https://github.com/cilium/cilium/blob/45689ece6f242c4c2546a3f32ddd80f6464f6e24/test/helpers/utils.go#L567-L572

pchaigno commented on test/k8sT/Services.go:
```diff
@@ -1979,6 +1982,11 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
+		SkipItIf(func() bool { return helpers.DoesNotRunOn419OrLaterKernel() || helpers.RunsWithKubeProxyReplacement() },
+			"Tests ClusterIP with sessionAffinity from pods", func() {
+				testSessionAffinity(false, true)
+			})
```
AFAIU, we still need to fix/remove this test as it will currently never run. `helpers.DoesNotRunOn419OrLaterKernel() || helpers.RunsWithKubeProxyReplacement()` is literally `a || ¬a`.

brb replied:
Shall we nominate 4.19 for running with kube-proxy?

pchaigno commented on test/k8sT/Services.go:
```diff
@@ -2056,6 +2059,15 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
 
+		// SA needs 4.19. However, on the 4.19 job we run KPR. So to test SA on
+		// bpf_lxc with kube-proxy we need to disable KRP.
+		SkipItIf(helpers.DoesNotRunOn419Kernel, "Tests kube-proxy ClusterIP with sessionAffinity from pods", func() {
```
```suggestion
		SkipItIf(func() bool {
			return helpers.RunsWithKubeProxy() && helpers.RunsOn419OrLaterKernel()
		}, "Tests kube-proxy ClusterIP with sessionAffinity from pods", func() {
```
And I think you can then make the comment shorter if you want :-)

pchaigno commented on test/k8sT/Services.go:
```diff
@@ -2056,6 +2059,15 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
+		SkipItIf(helpers.DoesNotRunOn419Kernel, "Tests kube-proxy ClusterIP with sessionAffinity from pods", func() {
+			DeployCiliumOptionsAndDNS(kubectl, ciliumFilename, map[string]string{
+				"kubeProxyReplacement": "disabled",
+			})
```
We then need to redeploy Cilium with KPR at the end of the test or the next test (`Tests IPv6 NodePort Services`) will run with kube-proxy handling requests instead of KPR.

joestringer reviewed (approved):
Docs side LGTM.

pchaigno reviewed (approved):
I reviewed only the two new commits.

pchaigno commented on test/k8sT/Services.go:
```diff
@@ -2056,6 +2059,22 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
 
+		// SA needs 4.19. However, on the 4.19 job we run KPR. So to test SA on
+		// bpf_lxc with kube-proxy we need to disable KRP.
+		SkipItIf(helpers.DoesNotRunOn419Kernel, "Tests ClusterIP with sessionAffinity", func() {
```
To fix the builds:
```suggestion
		SkipContextIf(helpers.DoesNotRunOn419Kernel, "Tests ClusterIP with sessionAffinity", func() {
```

brb replied:
Thanks!

joestringer reviewed (approved):
Ah thanks, I was going to propose something similar myself.

I note this is marked as a release-blocker but it doesn't seem to be actively being worked on (eg it's draft), so I am not planning to block the upcoming security releases on this PR.

brb committed:
test: Do not disable SessionAffinity on >= 4.19

The previous commit enabled the SessionAffinity feature when running
with KPR=off.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

pchaigno commented on test/helpers/kubectl.go:
```diff
@@ -2455,12 +2455,16 @@ func (kub *Kubectl) overwriteHelmOptions(options map[string]string) error {
 	if DoesNotRunOn419OrLaterKernel() {
 		addIfNotOverwritten(options, "kubeProxyReplacement", "disabled")
 		addIfNotOverwritten(options, "bpf.masquerade", "false")
-		addIfNotOverwritten(options, "sessionAffinity", "false")
```
:thinking: 

https://github.com/cilium/cilium/blob/dfa6b157e8e9484c65fd938b2b45a4f5f50c61f9/daemon/cmd/kube_proxy_replacement.go#L304-L306

:grin:

brb replied:
:cloud: :thinking: :cloud: :smile:

brb committed:
helm: Remove defaultSessionAffinity=true

The default prevented from disabling the option when installing w/o
upgradeCompatibility (1) or w/ upgradeCompatibility >= 1.8 (2).

This became problematic after we made it possible to have SA w/
KPR=disabled. On older kernels the cilium-agent fatals due to the
missing BPF LRU support.

The side-effect of this change is that for a previous installation the
session affinity will be disabled after the upgrade if the following
conditions are met:

1. The "kubeProxyReplacement" is set to "partial".
1. Either (1) or (2) was true for the previous installation.
2. A user did not explicitly set "sessionAffinity".

Signed-off-by: Martynas Pumputis <m@lambda.lt>

aanm reviewed (changes requested):
Can we clarify what are the consequences of a user running with session affinity: true and then set it to false?
